### DISCUSSION TITLE
DataGrid and Vertical Scrollbar and question about forcing a Slider change

### DISCUSSION THREAD

**Discussion by reo-1963** [AvaloniaUI/Avalonia#14970] [Q&A]
I'm using Avalonia and Avalonia.Controls.DataGrid 11.0.7 with MS Visual Studio 2022 and DotNet 7.0
I am able to programmatically load my DataGrid with data via the ItemsSource.
But I'd like to keep the selected item on screen when the User clicks a Refresh button even if the User has scrolled down the page when there are many pages of items.
Yes, I've tried the "datagrid.ScrollIntoView(item, col)" technique...
It does work. 
But it does not place the User at exactly the same place in the scroll view of the DataGrid that they were at.
And yes, keeping track of what column the User has sorted on and it's direction is a problem I have already handled
(and removed from the sample code because it just adds to the confusion).
I was able to get the vertical ScrollBar by walking through 3 tiers of "GetVisualChildren" until I found the item named "PART_VerticalScrollbar".
I am able to update the position of the ScrollBar to deep down it's scroll area.
But the ScrollBar does not move the visible elements of the DataGrid with it. The shown elements of the DataGrid stay at the top.

And my question is: how do I do I force the DataGrid to page and follow it's vertical ScrollBar? 

(When I run this code, and provided scrolling has happened once, and I am deep in the paging,
if I manually click the ScrollBar (that is moved halfway down the big list), the DataGrid jumps to that offset.
But I have to manually click the ScrollBar. So it feel like it is really close... but not quite)

Any help is appreciated.
Here is my sample code:

```
List <classToDisplay> listDisplay = new List<classToDisplay>();
//
// get a bunch of things to display and put into listDisplay
//
// Not pretty code to get the ScrollBar but it works
string sName = "PART_VerticalScrollbar";
ScrollBar verticalScrollBar = null; // datagrid.FindControl<ScrollBar>(sName); // returns null even if it is moved into the first Async area just under dScrollValue
IEnumerable<Visual> listKids = datagrid.GetVisualChildren();
foreach (Visual v in listKids)
{
	IEnumerable<Visual> listKids2 = v.GetVisualChildren();
	foreach (Visual v2 in listKids2)
	{
		IEnumerable<Visual> listKids3 = v2.GetVisualChildren();
		foreach (Visual v3 in listKids3)
		{
			if (v3.Name == sName)
			{
				verticalScrollBar = v3 as ScrollBar;
				break;
			}
		}
		if (verticalScrollBar != null)
			break;
	}
	if (verticalScrollBar != null)
		break;
}
double dScrollValue = 0;
Avalonia.Threading.Dispatcher.UIThread.InvokeAsync(async () =>
{
	if (verticalScrollBar != null)
	{ 
		dScrollValue = verticalScrollBar.Value;
	}
        classToDisplay selectedRowItem = datagrid.SelectedItem as classToDisplay;
        Guid? gSelected = selectedRowItem?.Guid;
   	// Load the sorted list into the grid
	datagrid.ItemsSource = listDisplay;
	if (gSelected != null)
	{
		//Avalonia.Threading.Dispatcher.UIThread.InvokeAsync(() =>
		//{
		// If the previously selected row can still be found, select it. If not, no harm, no foul
		classToDisplay cFound = listDisplay.Find(a => a.Guid == gSelected);
		datagrid.SelectedItem = cFound;

		// For the ScrollIntoView to work after the ItemsSource has just been set, it needs another Threading InvokeSync
		//datagrid.ScrollIntoView(cFound, null);
		} // );
	}
	if (verticalScrollBar != null)
	{
		Avalonia.Threading.Dispatcher.UIThread.InvokeAsync(async () =>
		{
			verticalScrollBar.Value = dScrollValue; // scrollbar moves but does not move the DataGrid with it.
		} );
	}
} );        
```

**Comment by timunie** [comment]
```cs
MyDataGrid.ScrollIntoView(myItem, MyDataGrid.Columns.First());
```

**Reply by reo-1963** [reply]
I said ScrollIntoView() will work...
But I am looking to keep the same amount of scrolling on the scrollbar when the ItemSource updates.

**Reply by timunie** [reply]
that's harder in a virtualized world. 🤔
